clash-lib/src/Clash/Rewrite/Types.hs

```diff
@@ -74,7 +75,9 @@ data RewriteStep
 data RewriteState extra
   = RewriteState
   { _transformCounter :: {-# UNPACK #-} !Int
-  -- ^ Number of applied transformations
+  -- ^ Total number of applied transformations
+  , _transformCounters :: Map String (Sum Word)
+  -- ^ Map that tracks how many times each transformation is applied
```

Comment:
Although I don't expect this dictionary to get large, `Map` should generally be avoided in favor of `HashMap`:

https://github.com/haskell-perf/dictionaries

Its memory footprint is way better and it only loses sligthly for very small dicts.

Reply:
Similarly, `String` should not be used as a key for `Map`s as its comparison overhead is truly awful. Now I might say `String` is truly awful in general, but I wouldn't want to start a flamewar :relieved:

Reply:
```haskell
'W' : 'h' : 'a' : 't' : '\'' : 's' : ' ' : 'w' : 'r' : 'o' : 'n' : 'g' : ' ' : 'w' : 'i' : 't' : 'h' : ' ' : 'S' : 't' : 'r' : 'i' : 'n' : 'g' : '?' : []
```